src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,638 @@
+//!    values are in the correct range, they are bit-encoded. (This is the same
+//!    as what happens in the `Sum` type.)
+//!    This means that instead of sending a field element in the range `[0,2^n)`,
+//!    we send `n` field elements representing the bit encoding. The validation
+//!    circuit can verify that all submitted "bits" are indeed either `0` or `1`.
+//! 4. The computed and submitted norms are treated very similar to the vector
+//!    entries, but they have a different number of bits, namely `2n-2`.
+//! 5. As the aggregation result is a pointwise sum of the client vectors,
+//!    the numbers no longer (semantically) lie in the range `[-1,1)`, and cannot
+//!    be represented by the same fixed point type as the input. Instead the
+//!    decoding happens directly into a vector of floats.
+//!
+//! ### Fixed point encoding
+//! Submissions consist of encoded fixed-point numbers in `[-1,1)` represented as
+//! field elements in `[0,2^n)`, where n is the number of bits the fixed-point
+//! representation has. Encoding and decoding is handled by the associated functions of the
+//! [`CompatibleFloat`](crate::flp::types::fixedpoint_l2::compatible_float::CompatibleFloat)
+//! trait. Semantically, the following function describes how a fixed-point value `x`
+//! in range `[-1,1)` is converted to a field integer:
+//! ```text
+//! enc : [-1,1) -> [0,2^n)
+//! enc(x) = 2^(n-1) * x + 2^(n-1)
+//! ```
+//! The inverse is:
+//! ```text
+//! dec : [0,2^n) -> [-1,1)
+//! dec(y) = (y - 2^(n-1)) * 2^(1-n)
+//! ```
+//! Note that these functions only make sense when interpreting all occuring
+//! numbers as real numbers. Since our signed fixed-point numbers are encoded as
+//! two's complement integers, the computation that happens in
+//! [`CompatibleFloat::to_field_integer`](crate::flp::types::fixedpoint_l2::compatible_float::CompatibleFloat::to_field_integer)
+//! is actually simpler.
+//!
+//! ### Norm computation
+//! The L2 norm of a vector xs of numbers in `[-1,1)` is given by:
+//! ```text
+//! norm(xs) = sqrt(sum_{x in xs} x^2)
+//! ```
+//! Instead of computing the norm, we make two simplifications:
+//! 1. We ignore the square root, which means that we are actually computing
+//!    the square of the norm.
+//! 2. Since we work with integers, fractional values should not occur in the
+//!    computation. This is done by working with a factor of `2^(2n-2)`.
+//! This means that what is actually computed in this type is the following:
+//! ```text
+//! our_norm(xs) = 2^(2n-2) * norm(xs)^2
+//! ```
+//! Given a vector ys of numbers in the field integer encoding (in `[0,2^n)`),
+//! this gives the following equation:
+//! ```text
+//! our_norm_on_encoded(ys) = our_norm([dec(y) for y in ys])
+//!                         = sum_{y in ys} y^2 - (2^n)*y + 2^(2n-2)
+//! ```
+//! The constant and linear terms in the sum appear because the decoding function
+//! is not linear but only affine.
```

Comment:
I think this is awkwardly phrased, because if the decode function _were_ linear and not affine, then there would still be the linear term in `our_norm_on_encoded`, right? As written, this implies that the function being affine is to blame for both terms.

Reply:
No, I would actually claim that the statement is correct as stated: Assume that the decoding function is linear, i.e.:

there is some `a : R`, such that:
```
dec : R -> R
dec(y) = a * y
```
then `our_norm_on_encoded`  would be (ignoring the `2^(2n-2)` factor which is irrelevant here, and which I am also going to explain in another comment)
```
our_norm_on_encoded = sum_{y in ys} dec(y)^2
                    = sum_{y in ys} (a*y)^2
```
where the summands have neither linear nor constant parts.